src/main/java/net/schmizz/sshj/transport/Proposal.java

```diff
@@ -127,13 +128,14 @@ public NegotiatedAlgorithms negotiate(Proposal other)
             throws TransportException {
         return new NegotiatedAlgorithms(
                 firstMatch(this.getKeyExchangeAlgorithms(), other.getKeyExchangeAlgorithms()),
-                allMatch(this.getSignatureAlgorithms(), other.getSignatureAlgorithms()),
+                firstMatch(this.getHostKeyAlgorithms(), other.getHostKeyAlgorithms()),
```

Comment:
I think we need to do a bit more magic here.... See RFC4253:

> kex_algorithms
         Key exchange algorithms were defined above.  The first
         algorithm MUST be the preferred (and guessed) algorithm.  If
         both sides make the same guess, that algorithm MUST be used.
         Otherwise, the following algorithm MUST be used to choose a key
         exchange method: Iterate over client's kex algorithms, one at a
         time.  Choose the first algorithm that satisfies the following
         conditions:
           +  the server also supports the algorithm,
           +  if the algorithm requires an encryption-capable host key,
            there is an encryption-capable algorithm on the server's
            server_host_key_algorithms that is also supported by the
            client, and
           +  if the algorithm requires a signature-capable host key,
            there is a signature-capable algorithm on the server's
            server_host_key_algorithms that is also supported by the
            client.
      If no algorithm satisfying all these conditions can be found, the
      connection fails, and both sides MUST disconnect.

> server_host_key_algorithms
         A name-list of the algorithms supported for the server host
         key.  The server lists the algorithms for which it has host
         keys; the client lists the algorithms that it is willing to
         accept.  There MAY be multiple host keys for a host, possibly
         with different algorithms.
         Some host keys may not support both signatures and encryption
         (this can be determined from the algorithm), and thus not all
         host keys are valid for all key exchange methods.
         Algorithm selection depends on whether the chosen key exchange
         algorithm requires a signature or an encryption-capable host
         key.  It MUST be possible to determine this from the public key
         algorithm name.  The first algorithm on the client's name-list
         that satisfies the requirements and is also supported by the
         server MUST be chosen.  If there is no such algorithm, both
         sides MUST disconnect.

Reply:
> I think we need to do a bit more magic here.... See RFC4253:
> 
> > kex_algorithms
> > Key exchange algorithms were defined above.  The first
> > algorithm MUST be the preferred (and guessed) algorithm.  If
> > both sides make the same guess, that algorithm MUST be used.
> > Otherwise, the following algorithm MUST be used to choose a key
> > exchange method: Iterate over client's kex algorithms, one at a
> > time.  Choose the first algorithm that satisfies the following
> > conditions:
> > +  the server also supports the algorithm,
> > +  if the algorithm requires an encryption-capable host key,
> > there is an encryption-capable algorithm on the server's
> > server_host_key_algorithms that is also supported by the
> > client, and
> > +  if the algorithm requires a signature-capable host key,
> > there is a signature-capable algorithm on the server's
> > server_host_key_algorithms that is also supported by the
> > client.
> > If no algorithm satisfying all these conditions can be found, the
> > connection fails, and both sides MUST disconnect.
> 
> > server_host_key_algorithms
> > A name-list of the algorithms supported for the server host
> > key.  The server lists the algorithms for which it has host
> > keys; the client lists the algorithms that it is willing to
> > accept.  There MAY be multiple host keys for a host, possibly
> > with different algorithms.
> > Some host keys may not support both signatures and encryption
> > (this can be determined from the algorithm), and thus not all
> > host keys are valid for all key exchange methods.
> > Algorithm selection depends on whether the chosen key exchange
> > algorithm requires a signature or an encryption-capable host
> > key.  It MUST be possible to determine this from the public key
> > algorithm name.  The first algorithm on the client's name-list
> > that satisfies the requirements and is also supported by the
> > server MUST be chosen.  If there is no such algorithm, both
> > sides MUST disconnect.

I agree that's true, but does SSHJ even implement a kex_algorithm that is encryption-based? Based on a quick glance they all seem to be signature-based, which would mean that the current behavior would be spec-compliant in practice. What do you think?

Reply:
Yes, but _in practice_ does not mean it will be spec compliant forever. Let's try to adhere to the spec and validate. We might get an encryption based kex sooner or later...

Do you know which ones are encryption based?

Reply:
I don't know of any encryption-based KEX algorithm, but will ask some people to make sure I'm not missing anything here.

Reply:
I [asked for examples on StackExchange](https://security.stackexchange.com/questions/235276/examples-of-ssh-key-exchange-algorithms-requiring-encryption-capable-host-keys) but have not received any yet.

I also started reading the OpenSSH Portable source code and it seems that they also don't implement support for key exchanges that are not signature-based: In [kex_choose_conf](https://github.com/openssh/openssh-portable/blob/fec89f32a84fd0aa1afc81deec80a460cbaf451a/kex.c#L882), both [kex algorithm](https://github.com/openssh/openssh-portable/blob/fec89f32a84fd0aa1afc81deec80a460cbaf451a/kex.c#L819) and [host key algorithm](https://github.com/openssh/openssh-portable/blob/fec89f32a84fd0aa1afc81deec80a460cbaf451a/kex.c#L839) are independently determined by the equivalent of `firstMatch`.

Given that OpenSSH itself doesn't implement this part of the SSH spec and there don't seem to be any examples (not even the new post-quantum kex algorithm is one), I would say that this implementation of the abstract negotiation algorithm is fine as it is. If someone wanted to use an encryption-based kex algorithm with SSHJ, they likely would need to make other changes as well, including to the kex interface itself.

Reply:
👍 Will add a few small additional comments in the PR

Reply:
Thanks!